Add auxiliary network devices support

DmytroLinkin opened the pull request:
Auxiliary devices pool is created when auxNetDevice type is specified in plugin config file. Auxiliary devices discovered for each discovered network PCI device, meaning only auxiliary network devices with parent PCI device are supported.
New AuxTypes selector allows to filter auxiliary device by the type.  For ex., given following auxiliary devices list..
```
foo.bar.0
fancy.bar.0
foo.bar.1
foo.baz.0
```
.. and selector value ["bar"], first three devices will be added to the pool.

Auxiliary network devices accept selectors: vendors, devices, drivers, pfNames, rootDevices, linkTypes, isRdma.

*NOTE* patchset built on top of #441

paravmellanox commented on README.md:
```diff
@@ -201,6 +206,15 @@ This plugin creates device plugin endpoints based on the configurations given in
             }
+        },
+        {
+            "resourceName": "bf2_sf",
```
things like this break in next 4 months with bf3.

There is nothing special about SF of bf2 vs bf3.

Please remove bf2_ prefix.

paravmellanox replied:
@DmytroLinkin better to have more practical names that users can copy.
Otherwise they will copy bf2_sf and it will fail when they upgrade sw to new generation asic.

DmytroLinkin replied:
It would not fail because it is not HW/SW dependent. Resource name is given by the admin.

paravmellanox commented on README.md:
```diff
@@ -31,20 +31,21 @@
 ## Features
 
-- Handles SR-IOV capable/not-capable devices (NICs and Accelerators alike)
+- Handles SR-IOV capable/not-capable devices (NICs and Accelerators alike, Auxiliry network devices)
```
its incorrect spelling.

it should be auxiliary

DmytroLinkin replied:
Mistype

moshe010 commented on pkg/auxnetdevice/auxNetDeviceProvider.go:
```diff
@@ -0,0 +1,193 @@
+				if newDevice, err := NewAuxNetDevice(device, auxDev, ap.rFactory, rc); err == nil {
+					newAuxDevices = append(newAuxDevices, newDevice)
+				} else {
+					glog.Errorf("auxnetdevice GetDevices(): error creating new device: %q", err)
```
I think this should be change to Warning. Also you just print the error can we have an Id so we can understand which auxDev we failed to add?

DmytroLinkin replied:
Yes

moshe010 commented on pkg/auxnetdevice/auxNetDeviceProvider.go:
```diff
@@ -0,0 +1,193 @@
+				}
+			}
+		} else {
+			glog.Errorf("auxnetdevice GetDevices(): error getting auxnetdevices from device %s: %q",
```
also should be change to Warning

DmytroLinkin replied:
Yes

moshe010 commented on pkg/auxnetdevice/auxNetDeviceProvider.go:
```diff
@@ -0,0 +1,193 @@
+		// discover auxiliary device names
+		auxDevs, err := utils.GetAuxNetDevicesFromPci(device.Address)
+		if err == nil {
+			for _, auxDev := range auxDevs {
```
can auxDevs  be empty? if yes we should log a warning that there is no auxDevs  for device.Address

DmytroLinkin replied:
Yes

moshe010 commented on pkg/auxnetdevice/auxNetDeviceProvider.go:
```diff
@@ -0,0 +1,193 @@
+	maxVendorNameLen  = 20
+	maxProductNameLen = 40
+	classIDBaseInt    = 16
+	classIDBitSize    = 64
```
where did you get the constant from the kernel? maybe it worth to point that out

DmytroLinkin replied:
I've copied this lines from other provider. But for sure they aren't defined in the kernel.

moshe010 commented on pkg/auxnetdevice/auxNetDeviceProvider.go:
```diff
@@ -0,0 +1,193 @@
+		if devClass == int64(deviceCode) {
+			vendor := device.Vendor
+			vendorName := vendor.Name
+			if len(vendor.Name) > maxVendorNameLen {
```
can you move line 87 + 88 to aux utils? 
something link normalizeVendorName()

DmytroLinkin replied:
Other providers also do this. I'll move to pkg/utils

moshe010 commented on pkg/auxnetdevice/auxNetDeviceProvider.go:
```diff
@@ -0,0 +1,193 @@
+			}
+			product := device.Product
+			productName := product.Name
+			if len(product.Name) > maxProductNameLen {
```
same comment for product

DmytroLinkin replied:
Ok

moshe010 commented on pkg/auxnetdevice/auxNetDeviceProvider.go:
```diff
@@ -0,0 +1,193 @@
+			}
+			glog.Infof("auxnetdevice AddTargetDevices(): device found: %-12s\t%-12s\t%-20s\t%-40s", device.Address,
+				device.Class.ID, vendorName, productName)
+			// exclude netdevice in-use in host XXX ???
```
why the ??? are you not sure if to skip netdevice in use?
So for VF this is what we do so I think it fine just need to logf that we as skipping that device because it has default route

DmytroLinkin replied:
Ok

DmytroLinkin replied:
In case of aux.devs it is parent PCI device, which is not used in the pool.

moshe010 commented on pkg/auxnetdevice/auxNetDeviceProvider.go:
```diff
@@ -0,0 +1,193 @@
+		if len(nf.LinkTypes) > 1 {
+			glog.Warningf("Link type selector should have a single value.")
+		}
+		if selector, err := rf.GetSelector("linkTypes", nf.LinkTypes); err == nil {
```
Do we have SF for IB? maybe this selector should be add when we have such support

DmytroLinkin replied:
SF requires switchdev mode. I don't know if it is possible to have it for IB, apparently not.

DmytroLinkin replied:
I've checked:
```
# ls /sys/class/net/ib0/device/*\.*\.* -d
/sys/class/net/ib0/device/mlx5_core.rdma.4
```

moshe010 commented on pkg/auxnetdevice/auxNetDeviceProvider.go:
```diff
@@ -0,0 +1,193 @@
+		glog.Errorf("unable to convert SelectorObj to AuxNetDeviceSelectors")
+		return false
+	}
+	if len(nf.AuxTypes) == 0 { // XXX needed?
```
I don't understand the comment

DmytroLinkin replied:
Do we need to check if user had provided `auxTypes` in config?

moshe010 commented on pkg/devices/gen_net.go:
```diff
@@ -59,6 +59,17 @@ func NewGenNetDevice(dev *ghw.PCIDevice, deviceID string, dt types.DeviceType, i
+		if pfAddr, err = utils.GetSriovnetProvider().GetPfPciFromAux(deviceID); err != nil {
+			return nil, err
+		}
+		// Only SF auxiliary devices can have an index, for other (-1, err) returned. Ignore this error
```
why we ignore errors?

DmytroLinkin replied:
Because if it is not SF, then error will be returned.

adrianchiris replied:
if deviceID does not point to an SF shouldnt we fail ?

what is the scenario where one or more of the methods below fail but we want NewGenNetDevice to return with no errors ?

DmytroLinkin replied:
> if deviceID does not point to an SF shouldnt we fail ?
> 

PR is adding support for auxiliary network devices, not SFs only.

> what is the scenario where one or more of the methods below fail but we want NewGenNetDevice to return with no errors ?

GetSfIndexByAuxDev - when device is not SF (and also when `sfnum` file read/parsing failed)
GetNetDevicesFromAux - due to the same reasons as VFs, I believe

adrianchiris replied:
> PR is adding support for auxiliary network devices, not SFs only.

Yes, however ATM we dont have a use-case for non SF AuxNetDevice.

moreover if the auxiliray device is a SF and we fail to get SF Index then i think we do want to fail.

DmytroLinkin replied:
> > PR is adding support for auxiliary network devices, not SFs only.
> 
> Yes, however ATM we dont have a use-case for non SF AuxNetDevice.
> 

In such case, why had I added aux. devices support and not just SFs?)

> moreover if the auxiliray device is a SF and we fail to get SF Index then i think we do want to fail.

It will also fail if no `sfnum` file presented, because it is non-SF device. I don't want plugin to fail in such case.

But Ok, I can make it fail and put a comment to review this line in the future

moshe010 commented on pkg/utils/utils.go:
```diff
@@ -450,3 +454,19 @@ func HasDefaultRoute(pciAddr string) (bool, error) {
 }
+
+// GetAuxNetDevicesFromPci returns list of auxiliary devices names for the specified PCI address
+func GetAuxNetDevicesFromPci(pciAddr string) ([]string, error) {
```
why this is not moved to sriovnet?

DmytroLinkin replied:
I can add it to sriovnet

DmytroLinkin replied:
https://github.com/k8snetworkplumbingwg/sriovnet/pull/55

DmytroLinkin commented on pkg/utils/utils.go:
```diff
@@ -28,6 +28,10 @@ import (
 	sysBusPci = "/sys/bus/pci/devices"
+	// golangci-lint doesn't see it is used in the testing.go
+	//nolint: unused
+	sysBusAux = "/sys/bus/auxiliary/devices"
```
Unit tests doesn't use it directly, the same as `sysBusPci`. But golangci-lint complains only on `sysBusAux` and I have no idea why.

adrianchiris commented on pkg/auxnetdevice/auxNetResourcePool.go:
```diff
@@ -0,0 +1,65 @@
+
+type auxNetResourcePool struct {
+	*resources.ResourcePoolImpl
+	selectors *types.NetDeviceSelectors
```
types.AuxNetDeviceSelectors ?

adrianchiris replied:
also as far as i see, selectors is not used in resource pool.  (also in other resource pools) so might as well remove it

DmytroLinkin replied:
This explains why it is working with `types.NetDeviceSelectors`

DmytroLinkin commented on README.md:
```diff
@@ -228,26 +243,53 @@ The "deviceType" value determines which selectors are supported for that device.
 #### Common selectors
```
Ok

adrianchiris commented on pkg/utils/utils.go:
```diff
@@ -423,3 +431,69 @@ func GetPfEswitchMode(pciAddr string) (string, error) {
+}
+
+// GetAuxNetDevicesFromPci returns list of auxiliary devices names for the specified PCI address
+func GetAuxNetDevicesFromPci(pciAddr string) ([]string, error) {
```
so, are we going to use : https://github.com/k8snetworkplumbingwg/sriovnet/pull/55 ?

if yes, this should be deleted.

DmytroLinkin replied:
Good catch. I probably forgot about this file.

adrianchiris commented on pkg/auxnetdevice/auxNetDevice.go:
```diff
@@ -0,0 +1,90 @@
+		return nil, err
+	}
+
+	auxType := ""
```
can you move this logic to utils ? to me it seems to contain too low level details for it to be here. 
(or just put it in separate method in this file)

also IMO for Aux device, if we fail to get its type (condition at L77 is false for some reason) we should fail creating it instead of creating it with no type.

DmytroLinkin replied:
Ok

adrianchiris commented on pkg/auxnetdevice/auxNetDeviceProvider.go:
```diff
@@ -0,0 +1,181 @@
+				}
+			}
+		} else {
+			glog.Warningf("auxnetdevice GetDevices(): error getting auxnetdevices from device %s: %q",
```
nit: auxNetDevice ? here and below at L77

DmytroLinkin replied:
Ok

adrianchiris commented on pkg/auxnetdevice/auxNetDeviceProvider.go:
```diff
@@ -0,0 +1,181 @@
+	if len(nf.AuxTypes) == 0 {
+		glog.Errorf("AuxTypes are not specified")
+		return false
+	}
```
do we want for now to support only SF Aux Device type ?

do you see today a use-case of an Aux device that is not SF ?

i was thinking for now to just ensure that AuxTypes has a single value of `"sf"` for now if there are no use-cases currently for other aux device types

DmytroLinkin replied:
I'm not aware about use-cases of non-SF aux. devices.
We can limit types to `"sf"` only, but for me it seems no problem that plugin will accept other types (assuming that error from getting SF index is ignored). There is just no known users for other types.

adrianchiris commented on pkg/auxnetdevice/auxNetDeviceProvider_test.go:
```diff
@@ -0,0 +1,192 @@
+			fs := &utils.FakeFilesystem{
+				Dirs: []string{
+					"sys/bus/auxiliary/devices",
+					"sys/bus/pci/devices/0000:00:00.1/foo.bar.0",
```
can we have examples here (throughout this file) based on real data where applicable ?

DmytroLinkin replied:
Ok

adrianchiris commented on pkg/auxnetdevice/auxNetDeviceProvider_test.go:
```diff
@@ -0,0 +1,192 @@
+
+				ve := []string{"8086", "8086", "1111", "2222", "3333"}
+				de := []string{"abcd", "123a", "abcd", "2222", "1024"}
+				md := []string{"tata", "toto", "toto", "toto", "titi"}
```
can we have examples here (throughout this file) based on real data where applicable ?

DmytroLinkin replied:
Ok

adrianchiris commented on pkg/auxnetdevice/auxNetDevice_test.go:
```diff
@@ -0,0 +1,189 @@
+				fakeSriovnetProvider := mocks.SriovnetProvider{}
+				fakeSriovnetProvider.
+					On("GetUplinkRepresentorFromAux", "foo.bar.0").Return("net0", nil).
+					On("GetPfPciFromAux", "foo.bar.0").Return("0000:00:00.1", nil).
```
can we have examples here (throughout this file) based on real data where applicable ?

DmytroLinkin replied:
Ok

adrianchiris commented on README.md:
```diff
@@ -31,11 +31,13 @@
 
 The SR-IOV Network Device Plugin is Kubernetes device plugin for discovering and advertising networking resources in the
-form of SR-IOV virtual functions (VFs) and PCI physical functions (PFs) available on a Kubernetes host.
+form of SR-IOV virtual functions (VFs), Auxiliary network devices (in particular Scalable functions) and PCI physical
```
```
.... in the form of:
- SR-IOV virtual functions (VFs)
- PCI physical functions (PFs)
- Auxiliary network functions, in particular scalable functions (SF)

which are available on a kubernetes host
```

@DmytroLinkin  should we refer in this document to scalable functions or sub functions ?
one is vendor implementation and the other is the upstream term.

we should use upstream term throughout if possible.

adrianchiris replied:
once we define here the abberviation SF, we can use it throughout this document instead of Scalable functions WDYT ?

DmytroLinkin replied:
> ```
> ....
> ```
> 
> @DmytroLinkin should we refer in this document to scalable functions or sub functions ? one is vendor implementation and the other is the upstream term.
> 
> we should use upstream term throughout if possible.

Until know I was sure that "scalable" is upstream one. I will change it to "sub" everywhere.

adrianchiris commented on README.md:
```diff
@@ -57,6 +59,9 @@ To deploy workloads with SR-IOV VF or PCI PF, this plugin needs to work together
   - Host device CNI (for PCI PFs)
     - During Pod creation, plumbs the allocated network device to the Pods network namespace using device information given by the meta plugin
     - On Pod deletion, reset and release the allocated network device from the Pod
+  - Ovn-kubernetes (for SFs Auxiliary netdevices and Switchdev VFs)
```
This work is already present upstream ?

DmytroLinkin replied:
No :see_no_evil:
But we always can merge this PR after, right?

adrianchiris replied:
it can be a follwup PR to update doc once PR is merged in ovn-kubernetes :)

adrianchiris replied:
can you remove this section ? once OVN-Kubernetes PR lands this section can be updated

adrianchiris commented on README.md:
```diff
@@ -285,40 +317,41 @@ Usage of ./sriovdp:
+8. "linkTypes"    - The link type of the net device associated with the PCI device
 
-The "pfNames" and "rootDevices" selectors can be used to specify a list and/or range of VFs for a pool in the below format:
+If device functions supports indexing (like Netdevice Virtual functions or Auxiliary netdevice Scalable functions), then "pfNames" and "rootDevices" selectors can be used to specify a list and/or range of functions for a pool in the below format:
```
if we already define at the top Auxiliary netdevice Scalable functions as SF
then here you can just write:

`The "pfNames" and "rootDevices" selectors can be used to specify a list and/or range of VFs/SFs for a pool in the below format`

DmytroLinkin replied:
Ok

adrianchiris commented on docs/scalable-functions/README.md:
```diff
@@ -0,0 +1,4 @@
+Scalable function (SF) is a lightweight function that has a parent PCI function on which it is deployed. Scalable function doesn't have PCI bus of its own, but is using [Auxiliary Bus](https://www.kernel.org/doc/html/latest/driver-api/auxiliary_bus.html) instead.
+
+One of the implementation is [Mellanox/ScalableFunctions](https://github.com/Mellanox/scalablefunctions/blob/main/README.md).
+Mellanox Scalable Function has its own function capabilities and its own resources. This means a Scalable Function has its own dedicated queues(txq, rxq, cq, eq). These queues are neither shared nor stolen from the parent PCI function. There is no special support needed from system BIOS to use Mellanox Scalable Functions. Scalable Functions do not require enabling PCI SR-IOV and co-exist with PCI ...
```
can you add a section of how to configure SFs so they can be consumed by device plugin ?

also we should have a heading for the page e.g

# Scalable Functions

add info about it

## Configure Scalable Function with Nvidia NIC

add info how to configuration with example sriov dp config to consume the resource

DmytroLinkin replied:
> can you add a section of how to configure SFs so they can be consumed by device plugin ?

Isn't Mellanox wiki page enough? It is well described there.

> 
> also we should have a heading for the page e.g
> 
> # Scalable Functions
> add info about it
> 
> ## Configure Scalable Function with Nvidia NIC
> add info how to configuration with example sriov dp config to consume the resource

Ok

DmytroLinkin commented on pkg/factory/factory_test.go:
```diff
@@ -350,6 +350,76 @@ var _ = Describe("Factory", func() {
+				vendors := []string{"8086", "8086", "8086", "8086"}
+				codes := []string{"1111", "1111", "1111", "1111"}
+				drivers := []string{"foo", "foo", "fancy", "foo"}
+				deviceID := []string{"foo.bar.0", "foo.bar.1", "fancy.baz.0", "foo.baz.0"}
```
Ok

adrianchiris commented on pkg/auxnetdevice/auxNetDevice.go:
```diff
@@ -0,0 +1,90 @@
+				isRdma = true
+				infoProviders = append(infoProviders, infoprovider.NewRdmaInfoProvider(rdmaSpec))
+			} else {
+				glog.Warningf("RDMA resources for %s not found. Are RDMA modules loaded?", dev.Address)
```
dev.Address -> deviceID

adrianchiris commented on pkg/auxnetdevice/auxNetDevice.go:
```diff
@@ -0,0 +1,90 @@
+	infoProviders := make([]types.DeviceInfoProvider, 0)
+	infoProviders = append(infoProviders, rFactory.GetDefaultInfoProvider(dev.Address, driverName))
+	isRdma := false
+	nf, ok := rc.SelectorObj.(*types.NetDeviceSelectors)
```
`types.AuxNetDeviceSelectors` ?

adrianchiris replied:
without this RDMA is not working

adrianchiris commented on go.mod:
```diff
@@ -3,13 +3,13 @@ module github.com/k8snetworkplumbingwg/sriov-network-device-plugin
 require (
-	github.com/Mellanox/rdmamap v1.0.0
-	github.com/Mellanox/sriovnet v1.0.3
+	github.com/Mellanox/rdmamap v1.0.1-0.20220811152017-9f6303b40e3a
```
can you bump to release version : https://github.com/Mellanox/rdmamap/tree/v1.1.0 ?

adrianchiris commented:
@DmytroLinkin 

i tested out your PR and added some comments.

i was able to successfully spin up a pod with SF + RDMA resources

one issue i found is that the injected env variable is showing the parent PF instead of the allocated SF.


```
/ # env

KUBERNETES_SERVICE_PORT=443

KUBERNETES_PORT=tcp://10.96.0.1:443

HOSTNAME=testpod1

SHLVL=1

HOME=/root

PCIDEVICE_NVIDIA_COM_MLNX_SF=0000:03:00.0

TERM=xterm

KUBERNETES_PORT_443_TCP_ADDR=10.96.0.1

PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

KUBERNETES_PORT_443_TCP_PORT=443

KUBERNETES_PORT_443_TCP_PROTO=tcp

KUBERNETES_SERVICE_PORT_HTTPS=443

KUBERNETES_PORT_443_TCP=tcp://10.96.0.1:443

KUBERNETES_SERVICE_HOST=10.96.0.1

PWD=/

```

DmytroLinkin commented:
> @DmytroLinkin
> 
> i tested out your PR and added some comments.
> 
> i was able to successfully spin up a pod with SF + RDMA resources
> 
> one issue i found is that the injected env variable (PCIDEVICE_NVIDIA_COM_MLNX_SF) is showing the parent PF instead of the allocated SF.
> 
> ```
> / # env
> 
> KUBERNETES_SERVICE_PORT=443
> 
> KUBERNETES_PORT=tcp://10.96.0.1:443
> 
> HOSTNAME=testpod1
> 
> SHLVL=1
> 
> HOME=/root
> 
> PCIDEVICE_NVIDIA_COM_MLNX_SF=0000:03:00.0
> 
> TERM=xterm
> 
> KUBERNETES_PORT_443_TCP_ADDR=10.96.0.1
> 
> PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
> 
> KUBERNETES_PORT_443_TCP_PORT=443
> 
> KUBERNETES_PORT_443_TCP_PROTO=tcp
> 
> KUBERNETES_SERVICE_PORT_HTTPS=443
> 
> KUBERNETES_PORT_443_TCP=tcp://10.96.0.1:443
> 
> KUBERNETES_SERVICE_HOST=10.96.0.1
> 
> PWD=/
> ```

It is genericInfoProvider which returns pciAddress of the device. I changed it to return pciAddr/auxiliary

adrianchiris commented:
/retest

adrianchiris commented on README.md:
```diff
@@ -79,6 +88,7 @@ The following  NICs were tested with this implementation. However, other SR-IOV
 - Mellanox ConnectX-5® Ex
 - Mellanox ConnectX-6®
 - Mellanox ConnectX-6® Dx
+- Mellanox BlueField-2®
```
can you add a section about creating SFs in the Main README ?

something similar to what we have below (L95)

DmytroLinkin replied:
L95 leed to `docs/vf-setup.md`. Should I point to chapter at `docs/subfunctions/README.md`?

adrianchiris replied:
yes

adrianchiris commented:
minor nit on doc update.

otherwise LGTM.

I have re-triggered Mellanox CI.

adrianchiris commented:
@SchSeba @Eoghan1232 PATL

adrianchiris reviewed (approved):
lgtm !

Thanks for working on it and addressing my comments @DmytroLinkin

adrianchiris commented:
@SchSeba @Eoghan1232 PTAL, id like to try and get this merged soon.

i believe the PR Is in a good state and LGTM'd it.

Eoghan1232 commented on pkg/auxnetdevice/auxNetDeviceProvider.go:
```diff
@@ -0,0 +1,189 @@
+	return nil
+}
+
+//nolint:gocyclo
```
should this be here? 
why do we want to skip?

Eoghan1232 replied:
thanks for clarifying

Eoghan1232 commented on pkg/devices/gen_net.go:
```diff
@@ -31,29 +33,52 @@ type GenNetDevice struct {
+	var funcID int
+	var err error
+
+	//nolint: exhaustive
```
same question as above.

DmytroLinkin replied:
Used it to make `accelerator` case go to `default`. If linter tag is undesired I will add case explicitly.

Eoghan1232 reviewed (approved):
lgtm!
Apologies, it took a while to get around to review all of it!

SchSeba reviewed (approved):
Amazing work!
/lgtm

adrianchiris commented:
Thx for working on this one @DmytroLinkin  !
